### PR TITLE
[SPARK-37429][PYTHON][MLLIB] Inline annotations for pyspark.mllib.linalg.__init__.py

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

Inline annotations for `pyspark.mllib.linalg.__init__.py`

### Why are the changes needed?

Currently, there is type hint stub files  `pyspark.mllib.linalg.__init__.pyi` to show the expected types for functions, but we can also take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: python/pyspark/mllib/linalg/__init__.py [hunk 1]
```diff
@@ -309,17 +354,18 @@ class DenseVector(Vector):
     DenseVector([-1.0, -2.0])
     """
 
-    def __init__(self, ar):
+    def __init__(self, ar: Union[bytes, np.ndarray, Iterable[float]]):
```
Comment (zero323) [inline python/pyspark/mllib/linalg/__init__.py:357]: This is intentionally simplified compared to `pyi` implementation ‒ there seems to be no advantage of keeping original overloads.

File: python/pyspark/mllib/linalg/__init__.py [hunk 2]
```diff
@@ -1325,24 +1468,26 @@ def __repr__(self):
...
+    def __reduce__(self) -> Tuple[Type["SparseMatrix"], Tuple[int, int, bytes, bytes, bytes, int]]:
         return SparseMatrix, (
             self.numRows,
             self.numCols,
-            self.colPtrs.tostring(),
-            self.rowIndices.tostring(),
-            self.values.tostring(),
+            self.colPtrs.tostring(),  # type: ignore[attr-defined]
+            self.rowIndices.tostring(),  # type: ignore[attr-defined]
+            self.values.tostring(),  # type: ignore[attr-defined]
```
Comment (zero323) [inline python/pyspark/mllib/linalg/__init__.py:1486]: For some reason, mypy doesn't recognize `ndarray.tostring`.

**Comment by HyukjinKwon** [comment]
cc @WeichenXu123 .

If there are no reviews being done, I would be fine to go ahead as are. I will leave it to you @zero323 - I have less background on ML side ..

**Review by HyukjinKwon — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: hs_err_pid626354.log [hunk 3]
```diff
@@ -0,0 +1,1796 @@
+#
```
Comment (HyukjinKwon) [inline hs_err_pid626354.log]: Seems like a mistake here :-)

File: replay_pid626354.log [hunk 4]
```diff
@@ -0,0 +1,1727 @@
+JvmtiExport can_access_local_variables 0
```
Comment (HyukjinKwon) [inline replay_pid626354.log]: here too :-)

File: hs_err_pid626354.log (hunk 3 above)
Comment (HyukjinKwon) [inline hs_err_pid626354.log]: Seems like a mistake :-).

File: replay_pid626354.log (hunk 4 above)
Comment (HyukjinKwon) [inline replay_pid626354.log]: ditto :-)

File: hs_err_pid626354.log (hunk 3 above)
Comment (zero323) [reply hs_err_pid626354.log]: Oh... JVM crash :( Thanks!

**Comment by zero323** [comment]
Merged into master.

**Pull request closed without merging**